### PR TITLE
[Mellanox]the port index in port_config.ini should starts from 0

### PR DESCRIPTION
**Pull request by stephenxs**
**- What I did**
The port index in port_config.ini should starts from 0, otherwise it will cause xcvrd crash.
```
Feb 11 21:06:01.716349 arc-switch1004 INFO pmon#supervisord: xcvrd SFP index 32 out of range (0-31)
Feb 11 21:06:01.716613 arc-switch1004 INFO pmon#supervisord: xcvrd Traceback (most recent call last):
Feb 11 21:06:01.716720 arc-switch1004 INFO pmon#supervisord: xcvrd   File "/usr/bin/xcvrd", line 1083, in <module>
Feb 11 21:06:01.716805 arc-switch1004 INFO pmon#supervisord: xcvrd     main()
Feb 11 21:06:01.716886 arc-switch1004 INFO pmon#supervisord: xcvrd   File "/usr/bin/xcvrd", line 1080, in main
Feb 11 21:06:01.716963 arc-switch1004 INFO pmon#supervisord: xcvrd     xcvrd.run()
Feb 11 21:06:01.717039 arc-switch1004 INFO pmon#supervisord: xcvrd   File "/usr/bin/xcvrd", line 1044, in run
Feb 11 21:06:01.717116 arc-switch1004 INFO pmon#supervisord: xcvrd     self.init()
Feb 11 21:06:01.717191 arc-switch1004 INFO pmon#supervisord: xcvrd   File "/usr/bin/xcvrd", line 1028, in init
Feb 11 21:06:01.717266 arc-switch1004 INFO pmon#supervisord: xcvrd     post_port_sfp_dom_info_to_db(is_warm_start, self.stop_event)
Feb 11 21:06:01.717343 arc-switch1004 INFO pmon#supervisord: xcvrd   File "/usr/bin/xcvrd", line 387, in post_port_sfp_dom_info_to_db
Feb 11 21:06:01.717421 arc-switch1004 INFO pmon#supervisord: xcvrd     post_port_sfp_info_to_db(logical_port_name, int_tbl, transceiver_dict, stop_event)
Feb 11 21:06:01.717498 arc-switch1004 INFO pmon#supervisord: xcvrd   File "/usr/bin/xcvrd", line 223, in post_port_sfp_info_to_db
Feb 11 21:06:01.717573 arc-switch1004 INFO pmon#supervisord: xcvrd     if not _wrapper_get_presence(physical_port):
Feb 11 21:06:01.717665 arc-switch1004 INFO pmon#supervisord: xcvrd   File "/usr/bin/xcvrd", line 123, in _wrapper_get_presence
Feb 11 21:06:01.717740 arc-switch1004 INFO pmon#supervisord: xcvrd     return platform_chassis.get_sfp(physical_port).get_presence()
Feb 11 21:06:01.717815 arc-switch1004 INFO pmon#supervisord: xcvrd AttributeError: 'NoneType' object has no attribute 'get_presence'
```

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-D48C8/port_config.ini [hunk 1]
```diff
@@ -1,57 +1,57 @@
...
+Ethernet76      76,77                  50000        etp20a      19
+Ethernet78      78,79                  50000        etp20b      19
+Ethernet80      80,81                  50000        etp21a      20
+Ethernet82      82,83                  50000        etp21b      20
+Ethernet84      84,85                  50000        etp22a      21
+Ethernet86      86,87                  50000        etp22b      21
+Ethernet88      88,89,90,91           100000        etp23       22
+Ethernet92      92,93,94,95           100000        etp24       23
+Ethernet96      96,97,98,99           100000        etp25       24
+Ethernet100     100,101,102,103        50000        etp26       25
```
Comment (prsunny) [inline device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-D48C8/port_config.ini]: Speed and lane info doesn't match here. Is it a typo?

**Comment by zhenggen-xu** [comment]
What is the definition of the index here? I saw most of the implementation was "front pannel" port number. In such case, it should be "1" based, and we need fix the other side instead.

File: device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-D48C8/port_config.ini (hunk 1 above)
Comment (stephenxs) [reply device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-D48C8/port_config.ini]: Nice catch.
Fixed.

**Comment by stephenxs** [comment]
> What is the definition of the index here? I saw most of the implementation was "front pannel" port number. In such case, it should be "1" based, and we need fix the other side instead.

Hi @zhenggen-xu,
AFAIK it should be 0 based otherwise it will fail the xcvrd.
Could you please provide an example of a "1"-based port_config.ini? I'd like to check. 
Thansk.

File: device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-D48C8/port_config.ini [hunk 2]
```diff
@@ -1,57 +1,57 @@
...
-Ethernet110     110,111                50000        etp28b      28
-Ethernet112     112,113                50000        etp29a      29
-Ethernet114     114,115                50000        etp29b      29
-Ethernet116     116,117                50000        etp30a      30
-Ethernet118     118,119                50000        etp30b      30
-Ethernet120     120,121                50000        etp31a      31
-Ethernet122     122,123                50000        etp31b      31
-Ethernet124     124,125                50000        etp32a      32
-Ethernet126     126,127                50000        etp32b      32
+Ethernet0       0,1                    50000        etp1a       0
```
Comment (prsunny) [inline device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-D48C8/port_config.ini]: Can you check if you need to make a corresponding change here ? https://github.com/Azure/sonic-mgmt/blob/master/ansible/library/minigraph_facts.py#L476

**Comment by prsunny** [comment]
Looks like this change is specific to fix a platform driver and would like Joe/Guohan to comment. Approving from my side!

File: device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-D48C8/port_config.ini (hunk 2 above)
Comment (stephenxs) [reply device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-D48C8/port_config.ini]: Hi Prince,
The purpose of this code is to generate a mapping from port alias to name. It doesn't reference the "port index" column. So no need to change it.

File: device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-C28D8/port_config.ini [hunk 3]
```diff
@@ -1,37 +1,37 @@
 # name          lanes                  speed        alias       index
```
Comment (nazariig) [inline device/mellanox/x86_64-mlnx_msn2700-r0/Mellanox-SN2700-C28D8/port_config.ini]: @stephenxs `# name         lanes              alias     index    speed    fec`

**Comment by zhenggen-xu** [comment]
> > What is the definition of the index here? I saw most of the implementation was "front pannel" port number. In such case, it should be "1" based, and we need fix the other side instead.
> 
> Hi @zhenggen-xu,
> AFAIK it should be 0 based otherwise it will fail the xcvrd.
> Could you please provide an example of a "1"-based port_config.ini? I'd like to check.
> Thanks.

https://github.com/Azure/sonic-buildimage/blob/master/device/celestica/x86_64-cel_seastone-r0/Celestica-DX010-C32/port_config.ini

My point is the xcvrd should handle both cases, it should probably use the port_config.ini as Source of Truth?

**Comment by stephenxs** [comment]
> > > What is the definition of the index here? I saw most of the implementation was "front pannel" port number. In such case, it should be "1" based, and we need fix the other side instead.
> > 
> > 
> > Hi @zhenggen-xu,
> > AFAIK it should be 0 based otherwise it will fail the xcvrd.
> > Could you please provide an example of a "1"-based port_config.ini? I'd like to check.
> > Thanks.
> 
> https://github.com/Azure/sonic-buildimage/blob/master/device/celestica/x86_64-cel_seastone-r0/Celestica-DX010-C32/port_config.ini
> 
> My point is the xcvrd should handle both cases, it should probably use the port_config.ini as Source of Truth?

Hi zhenggen,
Thank you for pointing out. 
Let's handle the issue on Mellanox-2700-D48C8 for now in this PR and consider a comprehensive solution in the future.

**Comment by zhenggen-xu** [comment]
> > > > What is the definition of the index here? I saw most of the implementation was "front pannel" port number. In such case, it should be "1" based, and we need fix the other side instead.
> > > 
> > > 
> > > Hi @zhenggen-xu,
> > > AFAIK it should be 0 based otherwise it will fail the xcvrd.
> > > Could you please provide an example of a "1"-based port_config.ini? I'd like to check.
> > > Thanks.
> > 
> > 
> > https://github.com/Azure/sonic-buildimage/blob/master/device/celestica/x86_64-cel_seastone-r0/Celestica-DX010-C32/port_config.ini
> > My point is the xcvrd should handle both cases, it should probably use the port_config.ini as Source of Truth?
> 
> Hi zhenggen,
> Thank you for pointing out.
> Let's handle the issue on Mellanox-2700-D48C8 for now in this PR and consider a comprehensive solution in the future.

If the indexes are presenting front pannel ports, to fix the error in xvrd, we need reload some of the plugins in your platform code in sonic_platform directory.  The changes are fine if the 0 based port index is what you are looking for.

**Comment by liat-grozovik** [comment]
retest vsimage please

**Comment by stephenxs** [comment]
Retest vsimage please

**Comment by stephenxs** [comment]
retest vsimage please

**Comment by liat-grozovik** [comment]
retest please

**Comment by liat-grozovik** [comment]
retest vsimage please

**Comment by stephenxs** [comment]
retest vsimage please

**Comment by stephenxs** [comment]
Hi @jleveque, could you please cherry-pick it to 201911? Thanks.

**Comment by qiluo-msft** [comment]
Fixed https://github.com/Azure/sonic-utilities/issues/588

**Comment by lguohan** [comment]
@zhenggen-xu is right. Check out the porting guide here.

https://github.com/Azure/SONiC/wiki/Porting-Guide

index: Physical index of port (optional - needed if ports do not begin with index zero or increment by one), When this column is missing the index starts from 0 and increase by 1 for each row. It is recommend to explicitly set this index column and make the index number match the labeling of the front panel ports. AKA: if the front panel ports are starting from 1, the index should start from 1. If the front panel ports start from 0, the index should start form 0. The break out ports from the same physical port should have same index.

**Pull request merged**